src/rust/engine/fs/src/lib.rs

```diff
@@ -482,72 +502,69 @@ impl PosixFS {
   }
 
   ///
-  /// Makes a Stat for path_for_stat relative to absolute_path_to_root.
+  /// Makes a Stat for path_to_stat relative to its containing directory.
   ///
   /// This method takes both a `FileType` and a getter for `Metadata` because on Unixes,
   /// directory walks cheaply return the `FileType` without extra syscalls, but other
   /// metadata requires additional syscall(s) to compute. We can avoid those calls for
   /// Dirs and Links.
   ///
   fn stat_internal<F>(
-    absolute_path_to_root: &Path,
-    path_for_stat: PathBuf,
+    path_to_stat: &Path,
     file_type: std::fs::FileType,
     compute_metadata: F,
   ) -> Result<Option<Stat>, io::Error>
   where
     F: FnOnce() -> Result<std::fs::Metadata, io::Error>,
   {
-    if !path_for_stat.is_relative() {
+    let Some(file_name) = path_to_stat.file_name() else {
       return Err(io::Error::new(
         io::ErrorKind::InvalidInput,
-        format!(
-          "Argument path_for_stat to PosixFS::stat must be relative path, got {path_for_stat:?}"
-        ),
+        "Argument path_to_stat to PosixFS::stat must have a file name.",
       ));
-    }
-    // TODO: Make this an instance method, and stop having to check this every call.
-    if !absolute_path_to_root.is_absolute() {
+    };
+    if !path_to_stat.is_absolute() {
       return Err(io::Error::new(
         io::ErrorKind::InvalidInput,
         format!(
-          "Argument absolute_path_to_root to PosixFS::stat must be absolute path, got {absolute_path_to_root:?}"
+          "Argument path_to_stat to PosixFS::stat must be absolute path, got {path_to_stat:?}"
         ),
       ));
     }
+    let path = file_name.to_owned().into();
     if file_type.is_symlink() {
       Ok(Some(Stat::Link(Link {
-        path: path_for_stat.clone(),
-        target: std::fs::read_link(absolute_path_to_root.join(path_for_stat))?,
+        path,
+        target: std::fs::read_link(path_to_stat)?,
       })))
     } else if file_type.is_file() {
       let is_executable = compute_metadata()?.permissions().mode() & 0o100 == 0o100;
       Ok(Some(Stat::File(File {
-        path: path_for_stat,
+        path,
         is_executable: is_executable,
       })))
     } else if file_type.is_dir() {
-      Ok(Some(Stat::Dir(Dir(path_for_stat))))
+      Ok(Some(Stat::Dir(Dir(path))))
     } else {
       Ok(None)
     }
   }
 
+  ///
+  /// Returns a Stat relative to its containing directory.
+  ///
   /// NB: This method is synchronous because it is used to stat all files in a directory as one
   /// blocking operation as part of `scandir_sync` (as recommended by the `tokio` documentation, to
   /// avoid many small spawned tasks).
-  pub fn stat_sync(&self, relative_path: PathBuf) -> Result<Option<Stat>, io::Error> {
-    let abs_path = self.root.0.join(&relative_path);
+  ///
+  pub fn stat_sync(&self, relative_path: &Path) -> Result<Option<Stat>, io::Error> {
```

Comment:
Should this (debug) assert or error if `relative_path` is not in fact a relative path?